Name: <redacted> <redacted>

SPIRE ID: 33832362

Github repo: https://github.com/<redacted>/assignment-5-judging-flowers-<redacted>

Committ ID: 0dc4a7a55423b09b1835f45f623aaaef4b266d43

In [2]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

# Part 1: Load the dataset

In [3]:
# Load the dataset (load remotely or locally)
data = pd.read_csv('iris.csv')

# Output the first 15 rows of the data
print(data.head(15))

# Display a summary of the table information (number of datapoints, etc.)
print(data.info())


    sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14           5.8         

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

The data contains 5 columns of sepal_length, sepal_width, petal_length, petal_width, and species. The features are the first four and the label is the last one of species. The first 2 columns are about the sepal which is the outer parts of the flower that enclose a developing bud, and we track its length and width. In the third and fourth columns we trach the petal which is a modified leaf that surrounds the reproductive parts of a flower and we also get its length and width. In the last column we have our label which is the species of flower that is in by category in species. The mapping of our label to classes is from species to three distinct species of Iris flowers: setosa, versicolor, and virginica

# Part 2: Split the dataset into train and test

In [4]:
# Take the dataset and split it into our features (X) and label (y)
X = data[['sepal_length', 'sepal_width', 'petal_length', 'petal_width']]
y = data['species']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)


# Part 3: Logistic Regression

In [5]:
# i. Use sklearn to train a LogisticRegression model on the training set
model = LogisticRegression(max_iter=200)
model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_data_point = pd.DataFrame([[5.1, 3.5, 1.4, 0.2]], columns=['sepal_length', 'sepal_width', 'petal_length', 'petal_width'])
predicted_probabilities = model.predict_proba(sample_data_point)

# Predict probabilities for each possible class
predicted_probabilities = model.predict_proba(sample_data_point)

# Display the predicted probabilities
class_labels = model.classes_  # Get the order of class labels
probability_df = pd.DataFrame(predicted_probabilities, columns=class_labels)

print("Predicted Probabilities for the sample data point:")
print(probability_df)
print()

# iii. Report on the score for Logistic regression model, what does the score measure?
accuracy = model.score(X_test, y_test)
print("Accuracy of the Logistic Regression model:", accuracy)
print()
# The score (accuracy) measures the proportion of correct predictions made by the model out of the total number of predictions made. 
# Since it is 1.0 it means the model made correct predictions for all samples in the test set, indicating perfect accuracy.


# iv. Extract the coefficents and intercepts for the boundary line(s)

# Extract coefficients and intercept
coefficients = model.coef_
intercept = model.intercept_

# Display the results for each class
for i, class_label in enumerate(model.classes_):
    print(f"Class: {class_label}")
    print(f"Coefficients: {coefficients[i]}")
    print(f"Intercept: {intercept[i]}")
    print()

cv_scores = cross_val_score(model, X, y, cv=5)  # 5-fold cross-validation
print("Cross-Validation Scores:", cv_scores)
print("Mean Cross-Validation Score:", cv_scores.mean())

Predicted Probabilities for the sample data point:
     setosa  versicolor     virginica
0  0.978916    0.021084  2.471808e-08

Accuracy of the Logistic Regression model: 1.0

Class: setosa
Coefficients: [-0.42830334  0.96729205 -2.44742526 -1.03780396]
Intercept: 9.543564860347782

Class: versicolor
Coefficients: [ 0.5130202  -0.22029888 -0.21604909 -0.84453741]
Intercept: 1.898872785007611

Class: virginica
Coefficients: [-0.08471686 -0.74699317  2.66347435  1.88234137]
Intercept: -11.442437645355573

Cross-Validation Scores: [0.96666667 1.         0.93333333 0.96666667 1.        ]
Mean Cross-Validation Score: 0.9733333333333334


sepal_length - (let's denote it as $x_1$)

sepal_width - (let's denote it as $x_2$)

petal_length - (let's denote it as $x_3$)

petal_width - (let's denote it as $x_4$)

Class: Setosa
$
\text{logit}(P(\text{setosa})) = -0.4283 \cdot x_1 + 0.9673 \cdot x_2 - 2.4474 \cdot x_3 - 1.0378 \cdot x_4 + 9.5436
$

Class: Versicolor
$
\text{logit}(P(\text{versicolor})) = 0.5130 \cdot x_1 - 0.2203 \cdot x_2 - 0.2160 \cdot x_3 - 0.8445 \cdot x_4 + 1.8989
$

Class: Virginica
$
\text{logit}(P(\text{virginica})) = -0.0847 \cdot x_1 - 0.7470 \cdot x_2 + 2.6635 \cdot x_3 + 1.8823 \cdot x_4 - 11.4424
$


# Part 4: Support Vector Machine

In [9]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svc_model = SVC(kernel='linear', probability=True)
svc_model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
predicted_probabilities = svc_model.predict_proba(sample_data_point)

# Display the predicted probabilities
class_labels = svc_model.classes_  # Get the order of class labels
probability_df = pd.DataFrame(predicted_probabilities, columns=class_labels)

print("Predicted Probabilities for the sample data point:")
print(probability_df)
print()

# iii. Report on the score for the SVM, what does the score measure?
accuracy = svc_model.score(X_test, y_test)
print("Accuracy of the Support Vector Classifier model on the test set:", accuracy)

#The score messures the same thing as in part 3 of prediction accuracy, and since it is 1.0, the SVM model made correct predictions for all instances in the test set. 

cv_scores = cross_val_score(svc_model, X, y, cv=5)  # 5-fold cross-validation
print("Cross-Validation Scores:", cv_scores)
print("Mean Cross-Validation Score:", cv_scores.mean())


Predicted Probabilities for the sample data point:
     setosa  versicolor  virginica
0  0.973458    0.016989   0.009554

Accuracy of the Support Vector Classifier model on the test set: 1.0
Cross-Validation Scores: [0.96666667 1.         0.96666667 0.96666667 1.        ]
Mean Cross-Validation Score: 0.9800000000000001


# Part 5: Neural Network

In [10]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlp_model = MLPClassifier(max_iter=1000)  # You can adjust the architecture
mlp_model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
# Predict probabilities for each possible class
predicted_probabilities = mlp_model.predict_proba(sample_data_point)

# Display the predicted probabilities
class_labels = mlp_model.classes_  # Get the order of class labels
probability_df = pd.DataFrame(predicted_probabilities, columns=class_labels)

print("Predicted Probabilities for the sample data point:")
print(probability_df)
print()

# iii. Report on the score for the Neural Network, what does the score measure?
accuracy = mlp_model.score(X_test, y_test)
print("Accuracy of the MLP Classifier model on the test set:", accuracy)
print()
#The score messures the same thing as in part 3 and 4 of prediction accuracy, and since it is 1.0, the Neural network made correct predictions for all instances in the test set. 

# iv: Experiment with different options for the neural network, report on your best configuration
mlp_model2 = MLPClassifier(hidden_layer_sizes=(1,), max_iter=1000, random_state=42)
mlp_model2.fit(X_train, y_train)

mlp_model3 = MLPClassifier(learning_rate='constant', learning_rate_init=0.1, hidden_layer_sizes=(100,), max_iter=1000, random_state=42)
mlp_model3.fit(X_train, y_train)

mlp_model4 = MLPClassifier(hidden_layer_sizes=(1,), max_iter=2000, random_state=42)
mlp_model4.fit(X_train, y_train)

accuracy2 = mlp_model2.score(X_test, y_test)
print("Accuracy of the MLP Classifier two model on the test set:", accuracy2)
print()

accuracy3 = mlp_model3.score(X_test, y_test)
print("Accuracy of the MLP Classifier three model on the test set:", accuracy3)
print()

accuracy4 = mlp_model4.score(X_test, y_test)
print("Accuracy of the MLP Classifier four model on the test set:", accuracy4)
print()

# Model 3 is the best where we set a constant learning rate and 100 hidden laters with a max iterations of 1000. It's accuracy is always 1.0.

# Using the third model (mlp_model3) for cross-validation since it has the better configuration
cv_scores = cross_val_score(mlp_model3, X, y, cv=5)  # 5-fold cross-validation
print("Cross-Validation Scores for the best model (Model 3):", cv_scores)
print("Mean Cross-Validation Score for the best model (Model 3):", cv_scores.mean())


Predicted Probabilities for the sample data point:
    setosa  versicolor     virginica
0  0.99943     0.00057  1.473703e-11

Accuracy of the MLP Classifier model on the test set: 1.0



/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (2000) reached and the optimization hasn't converged yet.
  warnings.warn(


Accuracy of the MLP Classifier two model on the test set: 0.8666666666666667

Accuracy of the MLP Classifier three model on the test set: 1.0

Accuracy of the MLP Classifier four model on the test set: 0.9333333333333333

Cross-Validation Scores for the best model (Model 3): [1.         1.         0.96666667 0.93333333 1.        ]
Mean Cross-Validation Score for the best model (Model 3): 0.9800000000000001


# Part 6: K-Nearest Neighbors

In [11]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
knn_model = KNeighborsClassifier(n_neighbors=5)
knn_model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
predicted_class = knn_model.predict(sample_data_point)
predicted_probabilities = knn_model.predict_proba(sample_data_point)
print("Predicted Class for the sample data point:", predicted_class[0])

# iii. Report on the score for kNN, what does the score measure?
class_labels = knn_model.classes_  # Get the class labels
probability_df = pd.DataFrame(predicted_probabilities, columns=class_labels)

print("Predicted Probabilities for the sample data point:")
print(probability_df)
print()

# Report on the accuracy of the KNN model
y_pred = knn_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)

print("Accuracy of the k-Neighbors Classifier model on the test set:", accuracy)

# The accuracy method typically measures the model's accuracy on the provided dataset, meaning that it correclt predicted each outcome for each flower
# It means that using KNN with a k=5 we correctly classify each point in our test set

cv_scores = cross_val_score(knn_model, X, y, cv=5)  # 5-fold cross-validation
print("Cross-Validation Scores:", cv_scores)
print("Mean Cross-Validation Score:", cv_scores.mean())



Predicted Class for the sample data point: setosa
Predicted Probabilities for the sample data point:
   setosa  versicolor  virginica
0     1.0         0.0        0.0

Accuracy of the k-Neighbors Classifier model on the test set: 1.0
Cross-Validation Scores: [0.96666667 1.         0.93333333 0.96666667 1.        ]
Mean Cross-Validation Score: 0.9733333333333334


# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

The results of the notebook were perfect for each of the methods and algorthims we ran. Everytime we ran, besides testing with the nueral networks, we ended with an accuracy score of 1.0 meaning that we correctly predicted the test set everytime, as well as with a sample data point. I think that the Nueral Network perfomred the best because it had the highest probabilty (besides KNN), to classify our sample data point. That means it was the most accurate, besides KNN who got every point right with a probability of 1. I think that is because neural networks are good for large data set and learning trends quickly, so they can predict with the most accuracy. I think KNN also did a really great job because these flowers tend to have the same characteristics, making them easy to clasify by their neighbors. I was suprised at how well all of these algorithms worked, and how easy they were to implement. Knowing how they work in conjunction with using these advanced libraries makes me alot more confident in my outputs. 

I also added the cross val scores for each of the method and SVM and Neural Networks both scored a .98. This means that they average out to have the highest score vals across 5 tests and perform the best over different sets of the data.